content/post/02-gitlab-alpinelinux-v4-api.md

```diff
@@ -0,0 +1,195 @@
+
+```sh
+_get_project_id() {
+	repo="$(echo "$1" \
+			| sed -e 's|https://||g' -e 's|%2F|.|' \
+			| tr '/' '.' )"
+
+	# Create cache directory if it doesn't exist
+	[ -d "$XDG_CACHE_HOME"/mkmr ] || mkdir -p "$XDG_CACHE_HOME"/mkmr
+
+	# The project ID is cached, read it out
+	if [ -f "$XDG_CACHE_HOME"/mkmr/"$repo" ]; then
+		cat "$XDG_CACHE_HOME"/mkmr/"$repo"
+		return 0
+	fi
+
+	# Call the GitLab API to see the id of the repo and write it to the cache
+	curl --silent "$1" | jq '.id // empty' | tee "$XDG_CACHE_HOME"/mkmr/"$repo"
+
+	[ -s "$XDG_CACHE_HOME"/mkmr/"$repo" ] || return 1
+}
+```
+
+The code above finds the project id via a request to the GitLab API which can be slow
+so we cache the id of the project so you don't need to call it every time you have to
+make a merge request.
+
+## Crafting the JSON payload
+
+The GitLab API takes a JSON payload for its requests, so we need to craft one in shell
+for it, which is easier than initially thought.
+
+This is the first part of the payload, and will always appear, regardless of whatever
+options the user has passed to the script when creating the merge requests
+
+```sh
+	# JSON payload that will be used to create the branch
+	BODY="{
+		\"id\": \"${origin_owner}%2F${origin_repo}\",
+		\"source_branch\": \"$(git rev-parse --abbrev-ref HEAD)\",
+		\"remove_source_branch\": true,"
+```
+
+The second part adds labels to the merge request if the user passed them via the --labels
+option. And also sets the description of the merge request, which the user can set by passing
+--description and/or passing --edit (edit with $EDITOR).
+
+> the A-upgrade label is automatically added if it detects an Alpine Linux upgrade commit
+> <repo>/<pkg>: upgrade to <pkgver>
+>
+> The A-add label is automatically added if it detects an Alpine Linux add commit
+> <repo>/<pkg>: new aport
+>
+> The A-move label is automatically added if it detects an Alpine Linux move commit
+> <newrepo>/<pkg>: move from <oldrepo>
+
+Note that we use awk to convert all newlines to literal \n, this is required and GitLab will
+translate them back from literal \n to newlines.
+
+```sh
+	# If we set labels then append it to body
+	if [ -n "$LABEL" ]; then
+		BODY="$BODY
+		\"labels\": \"$LABEL\","
+	fi
+
+	if [ -n "$DESCRIPTION" ]; then
+		BODY="$BODY
+		\"description\": \"$(printf "%s" "$DESCRIPTION" | awk '{printf  "%s\\n", $0}' )\","
+	fi
+```
+
+The last part sets the target_branch, which is where you want to merge the merge request into,
+most of the time it is the master branch, but in some cases you might want to merge into other
+branches with --target-branch. 
+
+> There is also logic specific to Alpine Linux for setting the target_branch to one of its
+> stable branches for backporting
+
+It also passes the title, which by default is the title of the only commit of the merge request
+or the user picks a commit from the multiple that is on the merge requests or pass the --title
+option to write their own.
```

Comment:
This sentence is kind of hard to grasp. Maybe something like this would be easier:

` ` ` 
If the branch only contains one commit deviating from the target branch it sets the title and description of the merge request to the commit's text. If there are multiple deviating commits it allows the user to pick one of the commits to use its text. Users can also pass the `--title`  option to the script to write their own title
` ` `